### PR TITLE
VDS-865 Assign user group to a customer

### PR DESCRIPTION
**Pull request by trueboroda**
Backend for this.

### CODE REVIEW

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        public MemberEntity Member { get; set; }
+
+        public ObservableCollection<DemoMemberTagEntity> Tags { get; set; }
+
+        #endregion
+
+        public virtual DemoTaggedMember ToModel(DemoTaggedMember taggedMember)
+        {
+            if (taggedMember == null)
+                throw new ArgumentNullException(nameof(taggedMember));
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs:29]: Add braces

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                taggedMember.Tags = Tags.Select(x => x.Tag).ToList();
+            }
+            
+            return taggedMember;
+        }
+
+        public virtual DemoTaggedMemberEntity FromModel(DemoTaggedMember taggedMember, PrimaryKeyResolvingMap pkMap)
+        {
+            if (taggedMember == null)
+                throw new ArgumentNullException(nameof(taggedMember));
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs:51]: Add braces

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            if (pkMap == null)
+                throw new ArgumentNullException(nameof(pkMap));
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs:56]: Add braces

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            }
+
+
+            return this;
+        }
+
+        public virtual void Patch(DemoTaggedMemberEntity taggedMemberEntity)
+        {
+            if (taggedMemberEntity == null)
+                throw new ArgumentNullException(nameof(taggedMemberEntity));
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs:83]: Add braces

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberSearchService.cs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+                        var sortInfos = criteria.SortInfos;
+                        if (sortInfos.IsNullOrEmpty())
+                        {
+                            sortInfos = new[] { new SortInfo { SortColumn = ReflectionUtility.GetPropertyName<DemoTaggedMember>(x => x.MemberId) } };
+                        }
+
+                        query = query.OrderBySortInfos(sortInfos).ThenBy(x => x.Id);
+
+                        result.TotalCount = query.Count();
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberSearchService.cs]: CountAsync

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            var taggedMember = (await _searchService.SearchTaggedMembersAsync(criteria)).Results.FirstOrDefault();
+
+            return Ok(taggedMember);
+        }
+
+
+        /// <summary>
+        /// PUT: api/demo/members/tagged
+        /// </summary>
+        [HttpPut]
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs]: Put should be used only for updating. In this case, would be better to use HttpPost attribute

**Review by pushnitsa — CHANGES_REQUESTED** [review]

**Comment by pushnitsa** [comment]
I guess we want to invalidate storefront cache after adding tags. Have you checked that case?

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Customer/IDemoTaggedMemberRepository.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
+using System.Linq;
+using System.Threading.Tasks;
+using VirtoCommerce.DemoSolutionFeaturesModule.Data.Models.Customer;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Data.Repositories.Customer
+{
+    public interface IDemoTaggedMemberRepository : IRepository
+    {
+        public IQueryable<DemoTaggedMemberEntity> TaggedMembers { get; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Repositories/Customer/IDemoTaggedMemberRepository.cs]: remove redundant access definition

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        [ProducesResponseType(typeof(DemoTaggedMember), StatusCodes.Status200OK)]
+        public async Task<ActionResult<DemoTaggedMember>> GetDemoTaggedMember(string id)
+        {
+            var criteria = new DemoTaggedMemberSearchCriteria
+            {
+                MemberId = id,
+                Take = 1
+            };
+
+            var taggedMember = (await _searchService.SearchTaggedMembersAsync(criteria)).Results.FirstOrDefault();
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs]: Why you don't use the simple GetByIdAsync method?

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        [Route("tagged")]
+        [ProducesResponseType(typeof(void), StatusCodes.Status204NoContent)]
+        public async Task<ActionResult> SaveTaggedMember([FromBody] DemoTaggedMember taggedMember)
+        {
+            try
+            {
+                await _taggedItemService.SaveChangesAsync(new[] { taggedMember });
+            }
+            // VP-4690: Handling concurrent update exception - e.g. adding 2 tagged for 1 entity
+            catch (DbUpdateException e) when (e.InnerException is Microsoft.Data.SqlClient.SqlException sqlException && (sqlException.Number == 2601 || sqlException.Number == 2627))
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs]: Is it necessary to cathing exceptions? Maybe would be better do not to produce them?

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+
+        /// <summary>
+        /// POST: api/demo/members/search
+        /// </summary>
+        [HttpPost]
+        [Route("search/tagged")]
+        [ProducesResponseType(typeof(DemoTaggedMemberSearchResult), StatusCodes.Status200OK)]
+        public async Task<ActionResult<DemoTaggedMemberSearchResult>> Search([FromBody] DemoTaggedMemberSearchCriteria criteria)
+        {
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs:64]: Check argument for a null

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs (hunk 8 above)
Comment (trueboroda) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs]: it is copy past.

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Customer/DemoTaggedMemberSearchCriteria.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+                {
+                    _memberIds = new[] { MemberId };
+                }
+                return _memberIds;
+            }
+            set
+            {
+                _memberIds = value;
+            }
+        }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Customer/DemoTaggedMemberSearchCriteria.cs]: Remove

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Customer/DemoTaggedMemberSearchResult.cs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Customer
+{
+    public class DemoTaggedMemberSearchResult: GenericSearchResult<DemoTaggedMember>
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Customer/DemoTaggedMemberSearchResult.cs]: Add space

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Caching/Customer/DemoTaggedMemberCacheRegion.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Primitives;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Customer;
+using VirtoCommerce.Platform.Core.Caching;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Data.Caching.Customer
+{
+    public class DemoTaggedMemberCacheRegion : CancellableCacheRegion<DemoTaggedMemberCacheRegion>
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Caching/Customer/DemoTaggedMemberCacheRegion.cs:10]: Try to make static

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Data.Models.Customer
+{
+    public class DemoTaggedMemberEntity : AuditableEntity
+    {
+        public DemoTaggedMemberEntity()
+        {
+            Tags = new NullCollection<DemoMemberTagEntity>();
+        }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs:15]: Add empty string

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+            taggedMember.Id = Id;
+
+            taggedMember.CreatedBy = CreatedBy;
+            taggedMember.CreatedDate = CreatedDate;
+            taggedMember.ModifiedBy = ModifiedBy;
+            taggedMember.ModifiedDate = ModifiedDate;
+
+            taggedMember.MemberId = MemberId;
+            if (!taggedMember.Tags.IsNullCollection())
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Models/Customer/DemoTaggedMemberEntity.cs:39]: Add empty string

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberSearchService.cs [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        private readonly Func<IDemoTaggedMemberRepository> _repositoryFactory;
+        private readonly IPlatformMemoryCache _platformMemoryCache;
+        private readonly IDemoTaggedMemberService _taggedMemberService;
+
+        public DemoTaggedMemberSearchService(Func<IDemoTaggedMemberRepository> repositoryFactory, IPlatformMemoryCache platformMemoryCache, IDemoTaggedMemberService taggedMemberService)
+        {
+            _repositoryFactory = repositoryFactory;
+            _platformMemoryCache = platformMemoryCache;
+            _taggedMemberService = taggedMemberService;
+        }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberSearchService.cs:27]: Add empty string

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberSearchService.cs [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        public async Task<DemoTaggedMemberSearchResult> SearchTaggedMembersAsync(DemoTaggedMemberSearchCriteria criteria)
+        {
+            var cacheKey = CacheKey.With(GetType(), nameof(SearchTaggedMembersAsync), criteria.GetCacheKey());
+            return await _platformMemoryCache.GetOrCreateExclusiveAsync(cacheKey,
+                async (cacheEntry) =>
+                {
+                    cacheEntry.AddExpirationToken(DemoTaggedMemberSearchCacheRegion.CreateChangeToken());
+                    var result = new DemoTaggedMemberSearchResult();
+
+                    using (var repository = _repositoryFactory())
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberSearchService.cs]: Meaningful name

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberService.cs [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+            ClearCache(taggedMembers);
+
+            await _eventPublisher.Publish(new DemoTaggedMemberChangedEvent(changedEntries));
+        }
+
+        public async Task<DemoTaggedMember[]> GetByIdsAsync(string[] ids)
+        {
+            var cacheKey = CacheKey.With(GetType(), nameof(GetByIdsAsync), string.Join("-", ids));
+
+            var result = await _platformMemoryCache.GetOrCreateExclusiveAsync(cacheKey,async (cacheEntry) =>
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Services/Customer/DemoTaggedMemberService.cs]: Add space before async

**Review by pushnitsa — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Caching/Customer/DemoTaggedMemberCacheRegion.cs (hunk 13 above)
Comment (trueboroda) [reply src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Caching/Customer/DemoTaggedMemberCacheRegion.cs:10]: The static class should be derived from the object.

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Customer/DemoTaggedMemberSearchCriteria.cs [hunk 19]
```diff
@@ -0,0 +1,12 @@
+using System;
+using VirtoCommerce.Platform.Core.Common;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Customer
+{
+    public class DemoTaggedMemberSearchCriteria : SearchCriteriaBase
+    {
+        public DateTime? ChangedFrom { get; set; }
+
+        public string[] Ids { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Core/Models/Customer/DemoTaggedMemberSearchCriteria.cs]: MemberIds?

File: src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Models.Customer;
+using VirtoCommerce.DemoSolutionFeaturesModule.Core.Services.Customer;
+
+namespace VirtoCommerce.DemoSolutionFeaturesModule.Web.Controllers.Api
+{
+    [Route("api/demo/members")]
+    [ApiController]
+    public class DemoCustomerController : Controller
```
Comment (pushnitsa) [inline src/VirtoCommerce.DemoSolutionFeaturesModule.Web/Controllers/Api/DemoCustomerController.cs:14]: Add Authorize attribute

**Review by pushnitsa — CHANGES_REQUESTED** [review]

**Review by pushnitsa — APPROVED** [review]

**Comment by trueboroda** [comment]
It will bring with this PR. 
 https://github.com/VirtoCommerce/vc-module-demo-features/pull/70

**Pull request closed without merging**